applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/element_finite_difference_utility.cpp

```diff
@@ -0,0 +1,258 @@
+                                                Matrix& rOutput,
+                                                ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        if ( rElement.GetProperties().Has(rDesignVariable) )
+        {
+
+            // define working variables
+            Vector RHS_unperturbed;
+            Vector RHS_perturbed;
+
+            // Compute RHS before perturbion
+            rElement.CalculateRightHandSide(RHS_unperturbed, rCurrentProcessInfo);
+
+            if ( (rOutput.size1() != 1) || (rOutput.size2() != RHS_unperturbed.size() ) )
+                rOutput.resize(1, RHS_unperturbed.size());
+
+            // Save property pointer
+            Properties::Pointer p_global_properties = rElement.pGetProperties();
+
+            // Create new property and assign it to the element
+            Properties::Pointer p_local_property(Kratos::make_shared<Properties>(Properties(*p_global_properties)));
+            rElement.SetProperties(p_local_property);
+
+            // perturb the design variable
+            const double current_property_value = rElement.GetProperties()[rDesignVariable];
+            p_local_property->SetValue(rDesignVariable, (current_property_value + rPertubationSize));
+
+            // Compute RHS after perturbation
+            rElement.CalculateRightHandSide(RHS_perturbed, rCurrentProcessInfo);
+
+            // Compute derivative of RHS w.r.t. design variable with finite differences
+            for(IndexType i = 0; i < RHS_perturbed.size(); ++i)
+                rOutput(0, i) = (RHS_perturbed[i] - RHS_unperturbed[i]) / rPertubationSize;
+
+            // Give element original properties back
+            rElement.SetProperties(p_global_properties);
+
+            //call one last time to make sure everything is as it was before TODO improve this..
+            rElement.CalculateRightHandSide(RHS_perturbed, rCurrentProcessInfo);
+        }
+        else
+            if ( (rOutput.size1() != 0) || (rOutput.size2() != 0) )
+                rOutput.resize(0,0,false);
+
+        KRATOS_CATCH("");
+    }
+
+    void ElementFiniteDifferenceUtility::CalculateRightHandSideDerivative(Element& rElement,
+                                                const array_1d_component_type& rDesignVariable,
+                                                Node<3>& rNode,
+                                                const double& rPertubationSize,
+                                                Vector& rOutput,
+                                                ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        if( rDesignVariable == SHAPE_X || rDesignVariable == SHAPE_Y || rDesignVariable == SHAPE_Z )
+        {
+            KRATOS_WARNING_IF("ElementFiniteDifferenceUtility::CalculateRightHandSideDerivative", OpenMPUtils::IsInParallel() != 0)
+                << "This function is not thread safe for shape derivatives!" << std::endl;
```

Comment:
thanks to the critical this function IS threadsafe!
Therfore this warning message is misleading
It should be sth like
"This function cannot be omp-parallelized, please do not call it from within a parallel section, this is inefficient" (my wording is far from optimal but I think you get what I mena :) )

Reply:
😃